### PR TITLE
[INFRA] Fix and create new workflow to compile delta-spark examples

### PR DESCRIPTION
**Pull request by scottsand-db**
#### Which Delta project/connector is this regarding?

- [ ] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [X] Other (INFRA)

## Description

Create new workflow `spark_examples_test.yaml` that compiles the `examples/scala/build.sbt` project.

Also fixes such compilation so that it uses the local jars (previously it was incorrectly hardcoded to Delta 3.0).

This requires running `publishM2` beforehand.

## How was this patch tested?

CI tests and specifically tested on a commit that uses a new `clusterBy` API in 3.2: https://github.com/delta-io/delta/pull/3012

## Does this PR introduce _any_ user-facing changes?

No.

### CODE REVIEW

File: .github/workflows/spark_examples_test.yaml [hunk 1]
```diff
@@ -0,0 +1,52 @@
+name: "Delta Spark Local Publishing and Examples Compilation"
+on: [push, pull_request]
+jobs:
+  test:
+    runs-on: ubuntu-20.04
+    strategy:
+      matrix:
+        # These Scala versions must match those in the build.sbt
+        scala: [2.12.18, 2.13.13]
```
Comment (vkorukanti) [inline .github/workflows/spark_examples_test.yaml:9]: is one not enough?

File: .github/workflows/spark_examples_test.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+          sudo apt install libedit-dev
+        if: steps.git-diff.outputs.diff
+      - name: Run Delta Spark Local Publishing and Examples Compilation
+        # examples/scala/build.sbt will compile against the local Delta relase version (e.g. 3.2.0-SNAPSHOT).
+        # Thus, we need to publishM2 first so those jars are locally accessible.
+        # We publish storage explicitly so that it is available for the Scala 2.13 build. As a java project
+        # it is typically only released when publishing for Scala 2.12.
+        run: |
+          build/sbt clean
+          build/sbt storage/publishM2
```
Comment (vkorukanti) [inline .github/workflows/spark_examples_test.yaml:52]: don't we need uniform jars too? also why not just use the `sparkGroup/publishM2`?

File: .github/workflows/spark_examples_test.yaml (hunk 1 above)
Comment (scottsand-db) [reply .github/workflows/spark_examples_test.yaml:9]: We run spark tests _tests_ for both scala 2.12 and 2.13 ... I see no reason not do use both scala versions just to _compile_ the examples.

If you feel strongly about this, let me know

File: .github/workflows/spark_examples_test.yaml (hunk 2 above)
Comment (scottsand-db) [reply .github/workflows/spark_examples_test.yaml:52]: the line below publishes all scala targets against scala 2.13 `build/sbt "++ $SCALA_VERSION publishM2"`

but that exludes publishing delta-storage, which is a java target, and which is only published when running `build/sbt "++ 2.12.18 publishM2"`

I added a comment for this on line 48, let me know if it's not clear enough

File: .github/workflows/spark_examples_test.yaml (hunk 1 above)
Comment (vkorukanti) [reply .github/workflows/spark_examples_test.yaml:9]: The number of CI jobs is going up. I guess we can revisit this as part of the test CI overhaul.

**Review by vkorukanti — APPROVED** [review]

**Pull request merged**